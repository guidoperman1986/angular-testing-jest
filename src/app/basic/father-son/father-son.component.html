<h5>Hijo</h5>

<code>{{ client | json }}</code>

<br /><br />

<button data-test="btn-id" class="btn btn-outline-primary" *ngIf="client" (click)="onChange(5)">Cambiar ID</button>
<button data-test="btn-delete" class="btn btn-outline-danger" *ngIf="client" (click)="onDelete()">Borrar Cliente</button>
