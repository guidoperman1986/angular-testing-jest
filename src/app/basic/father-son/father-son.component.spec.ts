import { ComponentFixture, TestBed } from '@angular/core/testing';
import { FatherSonComponent } from './father-son.component';


describe('FatherSonComponent', () => {
  let component: FatherSonComponent;
  let fixture: ComponentFixture<FatherSonComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ FatherSonComponent ]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(FatherSonComponent);
    component = fixture.componentInstance;
    compiled = fixture.nativeElement;
    fixture.detectChanges();
    jest.clearAllMocks();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  test('should match the snapshot', () => {
    expect(compiled).toMatchSnapshot();
  })

  test('should not appear buttons if there is no client', () => {
    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(0);
  })  

  test('should appear two buttons if there is client', () => {
    component.client = {id: 1, name: 'Juan'};

    fixture.detectChanges();

    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(2);
  })

  test('should emit onDeleteClient when delete button was pressed', ()=>{
    component.client = {id: 1, name: 'Juan'};
    fixture.detectChanges();

    jest.spyOn(component.onDeleteClient, 'emit');
    
    const btnDelete = compiled.querySelector('[data-test=btn-delete]');
    btnDelete?.dispatchEvent(new Event('click'));

    expect(component.onDeleteClient.emit).toHaveBeenCalled();

  })

  test('should emit onClientUpdated when change button was pressed', ()=>{
    component.client = {id: 1, name: 'Juan'};
    fixture.detectChanges();

    jest.spyOn(component.onClientUpdated, 'emit');
    
    const btnChangeId = compiled.querySelector('[data-test=btn-id]');
    btnChangeId?.dispatchEvent(new Event('click'));

    expect(component.onClientUpdated.emit).toHaveBeenCalledWith({id: 5, name: 'Juan'});

  })

  test('should emit onChangeClient with id if client exists', () => {
    jest.spyOn(component.onClientUpdated, 'emit');    
    component.onChange(10);
    expect(component.onClientUpdated.emit).not.toHaveBeenCalled();
    
    component.client = {id: 1, name: 'Juan'};
    fixture.detectChanges();
    component.onChange(10);
    expect(component.onClientUpdated.emit).toHaveBeenCalledWith({id: 10, name: 'Juan'});
  })



});



