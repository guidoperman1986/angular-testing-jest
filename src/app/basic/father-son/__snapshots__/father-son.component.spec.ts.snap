// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherSonComponent should match the snapshot 1`] = `
<div
  id="root1"
>
  <h5>
    Hijo
  </h5>
  <code>
    
  </code>
  <br />
  <br />
  
  
</div>
`;
